### ISSUE TITLE
Consideration of function "w2pers" in Neural Points module?  

### ISSUE THREAD

**Issue by MobiusLqm** [Xharlie/pointnerf#60]
Hi, Charlie, thanks for your awesome work of PointNeRF.  
But I had a question about the code In "w2pers" function of NeuralPoints.py. 
According to my understanding and your codes，"point_xyz" is the point position in reference view coordinates, which has shape(N,3), and "Campos" is the origin of target-view camera in reference view coordinates, "camrotc2w" is the transform matrix to transform points from target view coordinates to reference view coordinates(According  to  "__getitem__" function of dtu_dataset.py). So  my question is:
1. what's the meaning of point_xyz_shift?
2. I think the second line of codes below have the same function of "Point_xyz_shift@camrotc2w", do you want to transform the point from reference view coords to target view coords? but point position is a row vector, why didn't you transpose the Matrix?
3. what is the purpose of dividing x and y by z to get 'xper' and 'yper'? 

`

point_xyz_shift = point_xyz[None, ...] - campos[:, None, :]  # (1,N,3)
xyz = torch.sum(camrotc2w[:, None, :, :] *point_xyz_shift[:, :, :, None], dim=-2)  # same as  point_xyz_shift@camrotc2w??? 

xper = xyz[:, :, 0] / xyz[:, :, 2]
yper = xyz[:, :, 1] / xyz[:, :, 2]

`
Looking forward to your reply!
best regards

**Comment by MobiusLqm** [comment]
@Xharlie

**Comment by Xharlie** [comment]
hi we have developed everything in perspective coordinate first, so, if it is not used in the word coordinate, it might be a historical issue

**Comment by SeaBird-1995** [comment]
@MobiusLqm , have you understand the function of `w2pers`? I still very confused about this function.

**Comment by kaichen-z** [comment]
Dear [MobiusLqm],

It's the same thing. I was confused in the first place. But considering the size of two tensors (B, 3, 3) and (B, 1, 3). The operation proposed by the author is correct.

**Comment by Yi-yezi** [comment]
@SeaBird-Go Have you understood the function of w2pers?I think,maybe this line
`xyz = torch.sum(camrotc2w[:, None, :, :]* point_xyz_shift[:, :, :, None],dim=-2)`

it should be use a matrix of w2c instead of c2w?
